src/gridfs/download.rs

```diff
@@ -0,0 +1,181 @@
+    /// the contents to the `destination`.
+    pub async fn download_to_tokio_writer<T>(&self, id: Bson, destination: &mut T) -> Result<()>
+    where
+        T: tokio::io::AsyncWrite + std::marker::Unpin,
+    {
+        let options = FindOneOptions::builder()
+            .read_concern(self.read_concern().cloned())
+            .selection_criteria(self.selection_criteria().cloned())
+            .build();
+
+        let file = match self
+            .inner
+            .files
+            .find_one(doc! { "_id": &id }, options)
+            .await?
+        {
+            Some(fcd) => fcd,
+            None => {
+                return Err(ErrorKind::InvalidArgument {
+                    message: format!("couldn't find file with id {}", &id),
+                }
+                .into());
+            }
+        };
+        self.download_to_tokio_writer_common(file, destination)
+            .await
+    }
+
+    /// Downloads the contents of the stored file specified by `id` and writes
+    /// the contents to the `destination`.
+    pub async fn download_to_futures_0_3_writer<T>(
+        &self,
+        id: Bson,
+        destination: &mut T,
+    ) -> Result<()>
+    where
+        T: futures_util::io::AsyncWrite + std::marker::Unpin,
+    {
+        self.download_to_tokio_writer(id, &mut destination.compat_write())
+            .await
+    }
+
+    /// Downloads the contents of the stored file specified by `filename` and writes the contents to
+    /// the `destination`. If there are multiple files with the same filename, the `revision` in the
+    /// options provided is used to determine which one to download. If no `revision` is specified,
+    /// the most recent file with the given filename is chosen.
+    pub async fn download_to_tokio_writer_by_name<T>(
+        &self,
+        filename: impl AsRef<str>,
+        destination: &mut T,
+        options: impl Into<Option<GridFsDownloadByNameOptions>>,
+    ) -> Result<()>
+    where
+        T: tokio::io::AsyncWrite + std::marker::Unpin,
+    {
+        let revision = options.into().and_then(|opts| opts.revision).unwrap_or(-1);
+        let (sort, skip) = if revision >= 0 {
+            (1, revision)
+        } else {
+            (-1, -revision - 1)
+        };
+        let options = FindOneOptions::builder()
+            .sort(doc! { "uploadDate": sort })
+            .skip(skip as u64)
+            .read_concern(self.read_concern().cloned())
+            .selection_criteria(self.selection_criteria().cloned())
+            .build();
+
+        let file = match self
+            .inner
+            .files
+            .find_one(doc! { "filename": filename.as_ref() }, options)
+            .await?
+        {
+            Some(fcd) => fcd,
+            None => {
+                return Err(ErrorKind::InvalidArgument {
+                    message: format!("couldn't find file with name {}", &filename.as_ref()),
```

Comment:
This error case might be one that a user would want to handle programatically, so I think it could be worth introducing an enum case of some kind for `FileNotFound`. We could do this at the `ErrorKind` level (e.g. `ErrorKind::GridFsFileNotFound`), but given that we have a `GridFs` case, I think there's merit in folding them up all under that.

e.g.
```rust
enum ErrorKind {
    GridFs(GridFsErrorKind),
    // rest of cases
}

enum GridFsErrorKind {
    #[non_exhaustive]
    FileNotFound { filename: String },
    Other { message: String },
}
```
WDYT?

Reply:
Good idea, added an enum to model this. I ended up including a variant for each of the errors returned, but that approach might be worth revisiting if the list of possible errors becomes too bloated as the rest of the API is implemented.